File: ps/add-strings.cpp
class Solution {
public:
    string addStrings(string num1, string num2) {
        int i = num1.size() -1, j = num2.size()-1, added;
        bool carry = false;
        string res = "";
        while (true) {
            if (i < 0 && j < 0) break;
            if (j < 0) {
                added = num1[i] - '0';
            } else if (i < 0) {
                added = num2[j] - '0';
            } else {
                added = num1[i] - '0' + num2[j] - '0';
            }
            if (carry) added += 1;
            
            if (added >= 10) {
                res += added - 10 + '0';
                carry = true;
            } else {
                res += added + '0';
                carry = false;
            }
            i--;
            j--;
        }
        if (carry) res += "1";
        string rev = "";
        for (i=res.size()-1;i>=0;i--) rev += res[i];
        return rev;
    }
};

File: ps/add-strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        max_num = max(len(num1), len(num2))
        reversed_num1 = num1[::-1]
        reversed_num2 = num2[::-1]

        carry = False
        res = ''
        for idx in range(max_num):
            if len(reversed_num1) <= idx:
                reversed_num1 += '0'
            if len(reversed_num2) <= idx:
                reversed_num2 += '0'

            added_num = int(reversed_num1[idx]) + int(reversed_num2[idx]) # use ord('0') later.
            if carry:
                added_num += 1
                carry = False

            if added_num >= 10:
                added_num -= 10
                carry = True
            res += str(added_num)

        if carry:
            res += '1'

        return res[::-1]